cmd/thanos/store.go

```diff
@@ -154,6 +181,57 @@ func runStore(
 		s := grpc.NewServer(defaultGRPCServerOpts(logger, reg, tracer)...)
 		storepb.RegisterStoreServer(s, gs)
 
+		g.Add(func() error {
+			return errors.Wrap(s.Serve(l), "serve gRPC")
+		}, func(error) {
+			l.Close()
+		})
+	} else if s3Config.Validate() {
+		var bkt objstore.Bucket
```

Comment:
So what I was on about re: adding `Close` to the `Bucket` interface:
That doesn't actually make any sense, but the reason is that all this code below here is copypasta from above, save for the `gcsClient.Close()` on L142. So if we can figure out a way to abstract the rest of this common code and still be able to do that, this duplication can go away. It's a similar story in the other `cmd/*` files.

Does that make sense? Or am I off base?

Reply:
I can totally see your point. So either bucket needs to have `Close` method, but alternatively we can have small `closeFn` function:
```        
       var (
         bkt objstore.Bucket 
         closeFn = func(){}
       )

       if gcsBucket != "" {
                ....
        } else if s3Config.Validate() {
                bkt, err := s3.NewBucket(s3Config, reg)
 		if err != nil {
 			return errors.Wrap(err, "create s3 client")
 		}
       } 

       if bkt != nil {        
 		bkt = objstore.BucketWithMetrics(*s3Config.Bucket, bkt, reg)
 
 		bktStore, err := store.NewBucketStore(
 			logger,
 			reg,
 			bkt,
 			dataDir,
 			indexCacheSizeBytes,
 			chunkPoolSizeBytes,
 		)
 		if err != nil {
 			return errors.Wrap(err, "create bucket store")
 		}
 		ctx, cancel := context.WithCancel(context.Background())
 
 		g.Add(func() error {
                        defer  closeFn()
 			err := runutil.Repeat(3*time.Minute, ctx.Done(), func() error {
 				if err := bktStore.SyncBlocks(ctx); err != nil {
 					level.Warn(logger).Log("msg", "syncing blocks failed", "err", err)
 				}
 				peer.SetTimestamps(gs.TimeRange())
 				return nil
 			})
 
 			bktStore.Close()
 
 			return err
 		}, func() {
 			cancel()
 		})
           }
```

Reply:
Yes, just setting a `var close func()` or `var closer io.Closer` seems to solve this.